libmultilabel/linear/linear.py

```diff
@@ -475,6 +472,43 @@ def train_cost_sensitive_micro(y: sparse.csr_matrix, x: sparse.csr_matrix, optio
     return {'weights': np.asmatrix(weights), '-B': bias, 'threshold': 0}
 
 
+def train_binary_and_multiclass(y: sparse.csr_matrix, x: sparse.csr_matrix, options: str):
+    """Trains a linear model for binary and multi-class data.
+
+    Args:
+        y (sparse.csr_matrix): A 0/1 matrix with dimensions number of instances * number of classes.
+        x (sparse.csr_matrix): A matrix with dimensions number of instances * number of features.
+        options (str): The option string passed to liblinear.
+
+    Returns:
+        A model which can be used in predict_values.
+    """
+    x, options, bias = prepare_options(x, options)
+    num_labels = y.shape[1]
+    y = np.squeeze(y.nonzero()[1])
+
+    with silent_stderr():
+        model = train(y, x, options)
+
+    train_labels = model.get_labels()
+    ind = np.array(train_labels, dtype='int')
+    weights = np.zeros((x.shape[1], num_labels))
+    if num_labels == 2 and '-s 4' not in options:
+        # For binary classification, liblinear returns weights
+        # with shape (number of instances * 1) except '-s 4'.
+        w = np.ctypeslib.as_array(model.w, (x.shape[1], 1))
+        weights[:, ind[0]] = w[:, 0]
+        weights[:, ind[1]] = -w[:, 0]
+    else:
+        # Map label to the original index
+        w = np.ctypeslib.as_array(model.w, (x.shape[1], num_labels))
+        weights[:, ind] = w
+
```

Comment:
Add a comment:
For labels not appeared in training, assign thresholds to -inf so they won't be predicted.